teps/0037-simplify-metrics.md

```diff
@@ -0,0 +1,230 @@
+
+### Large metrics volume
+
+Currently, there is too much metrics per `TaskRun`/`PipelineRun`. According to [metrics.md](https://github.com/tektoncd/pipeline/blob/master/docs/metrics.md), the metrics count of a `PipelineRun` with `n` `TaskRun` is approximately `15*(n+1)+n`. Majority of metrics comes from histogram.
+The amount of metrics is huge while cluster is under stress. This amount of metrics will cause severe cluster performance degration if users have monitoring system running unthrottled.
+
+Large metrics volume may also cause metrics loss. Some monitoring system has limitation on metrics count.
+For example, the maximum number of Prometheus metrics that the agent can consume from the target is 10000 according to [Prometheus Metrics Limitation](https://docs.sysdig.com/en/limit-prometheus-metric-collection.html).
+Once metrics count reaches limitation, new metrics will be dropped and we cannot receive them anymore.
+
+### Fine granularity
+
+Currently, Tekton pipelines collect metrics at `TaskRun` and `PipelineRun` level. Users usually find it too noisy and want
+high-level and aggregated metrics.
+
+### Goals
+
+- Reduce the metrics volume
+- Allow users to configure the metrics granularity as they want
+- Simplify metrics so that users can see what they care about easily
+
+### Non-Goals
+
+- Change the behavior of existing metrics ingestor (Prometheus, Stackdriver etc)
+- Add support to a new metrics ingestor to handle a large amount of metrics
+- Build a new metrics ingestor to handle a large amount of metrics
+
+### Use Cases
+
+Cluster admin users are able to configure Tekton pipelines to produce high level metrics and focus on overall status to make sure Tekton works fine.
+While single users who care about statistics of individual `TaskRun` can still keep the fine-grained metrics.
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation.  The "Design Details" section below is for the real
+nitty-gritty.
+-->
+Coarse-grained metrics should be provided to satisfy the users' needs. This can be implemented in 2 ways: changing `TaskRun` level metrics to namespace level and changing metrics type. These 2 methods are mutually exclusive.
+If users care about individual `TaskRun`'s performance, they can set `duration_seconds` type from histogram to gauge. If users care about overall performance, they can collect metrics at namespace level.
+
+### Change TaskRun level metrics to namespace level
+
+We can add a `config-observability` option to switch between `TaskRun` level metrics and namespace level.
+`metrics.namespace-level` field indicates whether to aggregate metrics at namespace level. When set to `true`, it will remove `Task` and `TaskRun` label in metrics. Sizable reduction in metrics count can
```

Comment:
i wonder if we even want to aggregate at the namespace level? it seems much more reasonable than what we are currently doing, just wondering if there are any specific user requests for that - and i could imagine a situation where you have many many namespaces (e.g. i think @skaegi has described scenarios where namespaces are created and destroyed for each run?) and itd be more useful to aggregate these metrics across namespaces

Reply:
The metrics aggregated across namespaces and for namespace might be order of magnitude smaller than those at taskrun / pipelinerun level. I think we should have the ability to generate metrics at all three levels, and have a config option to define the maximum level of granularity to be produced: cross-namespace, per-namespace, per-item,
For backward compatibility we might have to keep the default to `per-item` for sometime, and we could switch to `per-namespace` or `cross-namespace` as default over time.